### DISCUSSION TITLE
Setting random seeds for randomBN

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#6] [Q&A] [answered]
I tried using the following to fix the seed, but it doesn't seem to work for `randomBN`

```
import random
random.seed(20)
```
How can I fix a seed so that I can re-generate the same BN with the same CPT each time?

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch , We have an internal random generator initialized with `gum.initRandom()` (https://pyagrum.readthedocs.io/en/1.6.1/miscellaneous.html#pyAgrum.initRandom) but it seems that `gum.randomBN()` does not use it ... :-(
We will have a look. Stay tuned! :-)

**Comment by phwuil** [comment]
Just to let you know the progress we made [here](https://gitlab.com/agrumery/aGrUM/-/tree/features/randomAndBNGenerator)  : 

- the aGrUm's generator was not completely used by some codes. Some times ago, we moved from rand() to [`std::mt19937`](https://www.geeksforgeeks.org/stdmt19937-class-in-cpp/) and this move was not complete. 

- the BN generator has been more or less rewritten and is much faster now. 

- But still we have to fix your main problem (setting random seed seems to work for basic random number generation but not for `randomBN`)

**Reply by kenneth-lee-ch** [reply]
@phwuil Thanks for the update!

**Comment by phwuil** [comment, marked as answer]
@kenneth-lee-ch   : Fixed !

![image](https://user-images.githubusercontent.com/5980636/225329535-f8aeed84-aa68-4ecb-9c80-243e2b6e29dc.png)

(tomorrow in pyAgrum-nightly , soon in 1.6.2)

**Reply by kenneth-lee-ch** [reply]
Also, thank you so much!

**Reply by kenneth-lee-ch** [reply]
@phwuil I just reinstall `pyargrum-nightly` and I ran into an error that didn't occur previously from the `randBN()` function

```
import pyAgrum as gum
graphnames=['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15']
gum.randomBN(n=16, names=graphnames, ratio_arc=3)
```

![Screenshot 2023-03-16 at 3 36 07 PM](https://user-images.githubusercontent.com/24616330/225733661-627753ad-0e33-48e3-ba1b-d069772776f0.png)

**Comment by phwuil** [comment]
Hum, that's logical but I do not know how to prevent that : the new randomBN use the names X{i} in topological order ... So if you try to force the same names (using graphnames), you may try to change the name "X10" of a node whereas another has the same name.

Now you can just use : `gum.randomBN(n=16,ratior_arc=3)`

![image](https://user-images.githubusercontent.com/5980636/225768855-63f1a0cc-8779-4fa3-9c94-b59851e16c94.png)

**Comment by phwuil** [comment]
PS- 16.2 will finally be 1.7.0 tomorrow

**Comment by phwuil** [comment]
PS2- Finally, I found a way to prevent these names collision. Will be in 1.7.0 tomorrow :-)

**Comment by kenneth-lee-ch** [comment]
Thanks! I look forward to the latest version!

**Reply by phwuil** [reply]
:crossed_fingers:  :-)

**Reply by phwuil** [reply]
1.7.0 is here !

**Reply by kenneth-lee-ch** [reply]
@phwuil  It seems like I can no longer generate binary variables by having `domain_size =2`? I tried using `domain_size =1`, but it throws an error.

![Screenshot 2023-03-17 at 3 48 40 PM](https://user-images.githubusercontent.com/24616330/226014138-3a857f61-30ad-4dfd-8858-34be53a3c4d6.png)
![Screenshot 2023-03-17 at 3 49 24 PM](https://user-images.githubusercontent.com/24616330/226014585-8170ec4a-22b1-4b60-8377-bc46173edfb3.png)

**Comment by phwuil** [comment]
![image](https://user-images.githubusercontent.com/5980636/226105316-8312c410-ae49-4151-a42b-aa5ec45e904f.png)

:-) 1.7.1 is here !

**Reply by phwuil** [reply]
.. and by the way, I forgot in this CHANGELOG that I added a `gum.lib.export.toFastBN` :-)

![image](https://user-images.githubusercontent.com/5980636/226123099-1ad786d5-8416-45d9-88c4-3eb5f20fd5d7.png)

**Comment by kenneth-lee-ch** [comment]
@phwuil When I use ` gum.initRandom(self.seed)` along with `gum.randomBN` inside a function of a class. It seems like I cannot reproduce the graph for some reason.

**Comment by phwuil** [comment]
@kenneth-lee-ch , it seems to work here :

![image](https://user-images.githubusercontent.com/5980636/230553100-f0182245-ac13-418c-8877-68348003e8ac.png)

**Reply by kenneth-lee-ch** [reply]
@phwuil I didn't use a notebook but just compiled a `.py` file. Will that make any difference?

**Comment by phwuil** [comment]
It should not and it looks okt :
```python
import pyAgrum as gum

class A:
  def f(self,seed):
    gum.initRandom(seed)
    return gum.randomBN(names="ABCDEFG")

a=A()
seeds=[5,4,0,5,4]
for seed in seeds:
  print("************")
  print(f"** {seed=}")
  print("************")
  bn=a.f(seed)
  print(f'  => {bn.parents("G")=}')
  print(bn.cpt("A"))
```


run :
![image](https://user-images.githubusercontent.com/5980636/230553859-f73f7947-2ccf-4f3a-8fa9-107e53c82c4f.png)

**Reply by phwuil** [reply]
can you share a minimal example of your problem ?

**Reply by kenneth-lee-ch** [reply]
I think it's a problem on my end. Thank you so much!
